### DISCUSSION TITLE
Purpose of the separate test_design in Test

### DISCUSSION THREAD

**Discussion by IAmMarcelJung** [FPGA-Research/FABulous#218] [Q&A] [answered]
What is the purpose of a separate test design in the Test directory? Even if the simulation and emulation commands are not yet perfectly integrated into the shell, the scripts `run_simulation.sh` and `run_emulation_sim.sh` could still point to the actual user design. It confused me a few times, that there is a separate test design, when I wanted to test the fabric with a user design. Is this just a leftover or is there reason for the separation?

**Comment by KelvinChung2000** [comment]
Simulation and emulation are different. I think the problem is what happens under simulation and under emulation. When we run under emulation, do we bypass something that makes it run faster, or vice versa?  

Based on my understanding of how simulation and emulation should behave, a simulation will combine the user design into the fabric, which gets optimised by the synthesis and then gets run. On the other hand, emulation will be the hardware setup, and then we load in the bitstream to see what the device is doing. 

But based on my definition, I think our current "simulation" is an emulation, and I don't know what emulation is doing. But they both exist for a reason...

**Reply by IAmMarcelJung** [reply]
Thanks for your quick answer! I think either I got your answer wrong or you got my question wrong. I did not want to ask about the difference between simulation and emulation, but just for the meaning of the separate test design, which is completely unrelated to the user design in the `user_design` directory. This exists both for the emulation and the simulation, at least for the two scripts `run_simulation.sh` and `run_emulation_sim.sh`. It also affects `build_test_design.sh` since this runs the FABulous flow for the so called test design. Sorry if I didn't make that clear in my initial question!

However I get your point with the differentiation between simulation and emulation. I always thought emulation in the context of FABulous would mean running a fabric on an acutal FPGA and a simulation is run from a testbench. But this might be a topic for another discussion.

**Reply by KelvinChung2000** [reply, marked as answer]
The rational behind `test` and `user_design` in two folder is because what you testing the fabric might not be the same as what run on the fabric. For example, we might want to run test that will exercise all the LUTs and wires within the fabric, so we implement a fabric size shift register which have no practical use. The `user_design` will be for what the user want to implement onto the fabric. Since the user have a few predefined workload want to run, they can optimise the fabric for those workload. 

The counter example just so happen being a good smoke test and user design.

**Reply by IAmMarcelJung** [reply]
Thanks, this made things clear to me! Then it makes absolute sense to have both the `test_design` and `user_design`!
